subworkflows/local/singular_demultiplex/main.nf

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env nextflow
+
+//
+// Demultiplex Singular Genomics bases data using sgdemux
+//
+
+include { SGDEMUX } from "../../../modules/nf-core/sgdemux/main"
+
+workflow SINGULAR_DEMULTIPLEX {
+    take:
+        // Input fastq's must be bgzipped for compatibility with sgdemux
+        // samplesheet.csv must be a two column csv = Sample_Barcode,Sample_ID)
+        ch_flowcell     // [[id:"", lane:""],samplesheet.csv, path/to/bases/files]
+
+    main:
+
+        // MODULE: sgdemux
+        // Demultiplex the bases files
+        SGDEMUX( ch_flowcell )
+
+        // Generate meta for each fastq
+        ch_fastq_with_meta = generate_fastq_meta(SGDEMUX.out.sample_fastq)
+
+    emit:
+        fastq                   = ch_fastq_with_meta
+        // sample_json             = BASES2FASTQ.out.sample_json TODO no json
+        // qc_report               = BASES2FASTQ.out.qc_report TODO no qc report
+        generated_run_manifest  = SGDEMUX.out.per_project_metrics // TODO
+        metrics                 = SGDEMUX.out.metrics // TODO 
+        unassigned              = SGDEMUX.out.most_frequent_unmatched // TODO 
+        versions                = SGDEMUX.out.versions
+}
+
+/*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    FUNCTIONS
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+*/
+
+// Add meta values to fastq channel
+def generate_fastq_meta(ch_reads) {
+    // Create a tuple with the meta.id and the fastq
+    ch_reads.transpose().map{
+        fc_meta, fastq ->
+        def meta = [
+            "id": fastq.getSimpleName().toString() - ~/_R[0-9].*$/,
+            "samplename": fastq.getSimpleName().toString() - ~/_R[0-9].*$/,
+            "readgroup": [:],
+            "fcid": fc_meta.id,
+            "lane": fc_meta.lane
+        ]
+        meta.readgroup = readgroup_from_fastq(fastq)
+        meta.readgroup.SM = meta.samplename
+
+        return [ meta , fastq ]
+    }
+    // Group by meta.id for PE samples
+    .groupTuple(by: [0])
+    // Add meta.single_end
+    .map {
+        meta, fastq ->
+        if (fastq.size() == 1){
+            meta.single_end = true
+        } else {
+            meta.single_end = false
+        }
+        return [ meta, fastq.flatten() ]
+    }
+}
+
+// https://github.com/nf-core/sarek/blob/7ba61bde8e4f3b1932118993c766ed33b5da465e/workflows/sarek.nf#L1014-L1040
+def readgroup_from_fastq(path) {
```

Comment:
Unused, remove

Reply:
I believe `readgroup_from_fastq()` is called in `generate_fastq_meta()`?